### ISSUE TITLE
Please support Chroma is a 8.9 billion parameter rectified flow transformer capable of generating images from text descriptions. Based on FLUX.1 [schnell] with heavy architectural modifications.

### ISSUE THREAD

**Issue by nitinmukesh** [mit-han-lab/nunchaku#167]
https://huggingface.co/lodestones/Chroma

**Comment by pcprinciple** [comment]
yeah, it's a fantastic model.
I‘m also hope the community can support this model. However, the model is still in training, so it may be a good idea to wait until the model is officially released before quantizing it.

**Comment by josephrocca** [comment]
> it may be a good idea to wait until the model is officially released before quantizing it

This issue is about supporting the model within the Nunchaku engine (and DeepCompressor), not quantizing it. There's no need to wait once it has support - every version of Chroma will be quantized by someone in the community. SVDQuant is more expensive than the average quantization method, but it's still pennies basically, and it's relatively easy to do.

Also, RE the title of this issue "with heavy architectural modifications" - the modifications aren't too heavy. Mainly replacing the per-transformer-block guidance layers with a much smaller distilled version, and masking the `encoder_hidden_states` (leaving one padding token at the end) - see this comment for more info:

* https://github.com/huggingface/diffusers/issues/11010#issuecomment-2741228637

**Comment by josephrocca** [comment]
@lmxyy It's possible that Nunchaku and DeepCompressor already "support" Chroma, if it's possible to pass an `attention_mask` to `NunchakuFluxTransformerBlocks`. Is that possible? If it is, then Chroma can be converted into a Schnell-compatible form like this:

```py
# docker run --rm -it --gpus all -v /home/USER/Desktop/chroma:/workspace -w /workspace pytorch/pytorch:2.6.0-cuda12.6-cudnn9-devel bash

# Create Schnell-compatible variant of Chroma by downloading both Chroma and Schnell safetensor files, and copying Chroma's matching weights over to Schnell. This works because lodestone *distilled* the guidance layers instead of completely pruning them, so we can actually just use Schnell's guidance stuff. This comes at the cost of bloating the model back to Schnell's original size.
CHROMA_VERSION="15"
apt-get update && apt-get install aria2 -y && pip3 install safetensors
cd /workspace
aria2c -x 16 -s 16 -o chroma.safetensors "https://huggingface.co/lodestones/Chroma/resolve/main/chroma-unlocked-v${CHROMA_VERSION}.safetensors?download=true"
aria2c -x 16 -s 16 -o flux1-schnell.safetensors "https://huggingface.co/black-forest-labs/FLUX.1-schnell/resolve/main/flux1-schnell.safetensors?download=true"
python3 -c '
from safetensors import safe_open
from safetensors.torch import save_file
with safe_open("/workspace/chroma.safetensors", framework="pt", device="cpu") as chroma, safe_open("/workspace/flux1-schnell.safetensors", framework="pt", device="cpu") as schnell:
    chroma_tensors = {key: chroma.get_tensor(key) for key in chroma.keys()}
    schnell_tensors = {key: schnell.get_tensor(key) for key in schnell.keys()}
matching_keys = set(chroma_tensors).intersection(schnell_tensors)
for key in matching_keys:
    schnell_tensors[key] = chroma_tensors[key]
save_file(schnell_tensors, "/workspace/chroma-schnell-compat.safetensors")
'
```

So then we only need to:
1. Mask the padding tokens of the T5 prompt (except the last padding token)
2. Feed in `zeros_like` of the CLIP embeddings

**Comment by lmxyy** [comment]
> [@lmxyy](https://github.com/lmxyy) It's possible that Nunchaku and DeepCompressor already "support" Chroma, if it's possible to pass an `attention_mask` to `NunchakuFluxTransformerBlocks`. Is that possible? If it is, then Chroma can be converted into a Schnell-compatible form like this:
> 
> # docker run --rm -it --gpus all -v $(pwd):/workspace -w /workspace pytorch/pytorch:2.6.0-cuda12.6-cudnn9-devel bash
> 
> # Create Schnell-compatible variant of Chroma by downloading both Chroma and Schnell safetensor files, and copying Chroma's matching weights over to Schnell. This works because lodestone *distilled* the guidance layers instead of completely pruning them, so we can actually just use Schnell's guidance stuff. This comes at the cost of bloating the model back to Schnell's original size (which is fine for now).
> CHROMA_VERSION="15"
> apt-get update && apt-get install aria2 -y && pip3 install safetensors
> cd /workspace
> aria2c -x 16 -s 16 -o chroma.safetensors "https://huggingface.co/lodestones/Chroma/resolve/main/chroma-unlocked-v${CHROMA_VERSION}.safetensors?download=true"
> aria2c -x 16 -s 16 -o flux1-schnell.safetensors "https://huggingface.co/black-forest-labs/FLUX.1-schnell/resolve/main/flux1-schnell.safetensors?download=true"
> python3 -c '
> from safetensors import safe_open
> from safetensors.torch import save_file
> with safe_open("/workspace/chroma.safetensors", framework="pt", device="cpu") as chroma, safe_open("/workspace/flux1-schnell.safetensors", framework="pt", device="cpu") as schnell:
>     chroma_tensors = {key: chroma.get_tensor(key) for key in chroma.keys()}
>     schnell_tensors = {key: schnell.get_tensor(key) for key in schnell.keys()}
> matching_keys = set(chroma_tensors).intersection(schnell_tensors)
> for key in matching_keys:
>     schnell_tensors[key] = chroma_tensors[key]
> save_file(schnell_tensors, "/workspace/chroma-schnell-compat.safetensors")
> '
> So then we only need to:
> 
> 1. Mask the padding tokens of the T5 prompt (except the last padding token), construct the `attention_mask` (for the joint txt+img transformer blocks only) [like this](https://github.com/lodestone-rock/flow/blob/a704bcf24df995563e68f4a4fdda985bacb7aad8/src/models/chroma/model.py#L223-L241)
> 2. Feed in `torch.zeros` for the CLIP embeddings (Chroma doesn't need/use the CLIP embeddings)

Should be possible

**Comment by josephrocca** [comment]
I kinda sorta got it working here in a very hacky manner:

* https://gist.github.com/josephrocca/dd339fc20c18dd48524c57b6e4005486

The only weird thing is that when using the quants, I need to set CFG to 1 for the last couple of steps or it burns the image. So I'm probably doing something wrong there.

**Comment by josephrocca** [comment]
Actually, the "set CFG to 1 for the last couple of steps" hack isn't needed anymore. I think it was due to uncond being unstable in earlier versions of Chroma. SVDQuant of v27 seems to work well in my initial testing.

**Comment by lmxyy** [comment]
@Aprilhuu Could you take a look? This model seems quite similar to FLUX and should be straightforward to support in both DeepCompressor and Nunchaku.

**Comment by Aprilhuu** [comment]
> [@Aprilhuu](https://github.com/Aprilhuu) Could you take a look? This model seems quite similar to FLUX and should be straightforward to support in both DeepCompressor and Nunchaku.

Ok I will take a look

**Comment by josephrocca** [comment]
@Aprilhuu In case it's helpful, references to the Chroma code for the key changes:

* Attention masking (leave one padding token after the prompt tokens): https://github.com/lodestone-rock/flow/blob/8ce37386953f6a8284612de7e845dafe86fddc3b/src/models/chroma/model.py#L223-L241
* Distilled modulation stuff: https://github.com/lodestone-rock/flow/blob/8ce37386953f6a8284612de7e845dafe86fddc3b/src/models/chroma/model.py#L195-L212

And safetensors file:

https://huggingface.co/lodestones/Chroma/resolve/main/chroma-unlocked-v27.safetensors?download=true

And arch diagram showing `distribute_modulations` stuff:

![Image](https://github.com/user-attachments/assets/1c26a1b5-24b5-4441-a3f2-a211e61ed5d6)

**Comment by sorasoras** [comment]
I am looking forward for support of chroma as well

**Comment by josephrocca** [comment]
@Aprilhuu @lmxyy just wondering if this is being actively worked on? This issue now has the highest thumbs up count of all feature requests on the Nunchaku repo 🚀 If diffusers support is a prerequisite and is blocking this, then please let me know since I might be able to help speed that up, assuming you're waiting for an external contributor to solve that.

For those who are impatient, I've uploaded some unofficial, possibly low-quality/buggy weights here. Be sure to read the usage steps since the weights are Chroma "pretending to be" Schnell:

* https://huggingface.co/rocca/chroma-nunchaku-test

**Comment by nitinmukesh** [comment]
I think major issue with support in diffusers was related to Chroma requiring masking/truncation of prompts. This was already taken care 
https://github.com/huggingface/diffusers/pull/11120

**Comment by nitinmukesh** [comment]
@josephrocca 

Thank you for sharing the weights. Are these specific to Comfy or can also be used with example code (just replacing the transformer)
https://github.com/mit-han-lab/nunchaku/blob/main/examples/flux.1-schnell.py

**Comment by josephrocca** [comment]
@Aprilhuu It's no thanks to me, but there's now a draft PR here:
* https://github.com/huggingface/diffusers/pull/11566

**Comment by Aprilhuu** [comment]
> [@Aprilhuu](https://github.com/Aprilhuu) There's a draft PR by [@hameerabbasi](https://github.com/hameerabbasi) taking shape here:
> 
> * [Chroma as a FLUX.1 variant huggingface/diffusers#11566](https://github.com/huggingface/diffusers/pull/11566)

@josephrocca I tried to hack through this draft PR but ran into a series of errors, even after switching to the new convert function. It also looks like the plan is to separate Chroma into its own class rather than keeping it as a Flux variant, so it seems like more structural changes are on the way. To avoid duplicating effort, I’ll probably wait until there’s a working PR. Lmk if you’ve had any luck getting it to work on your end.

**Comment by lawlietdream** [comment]
> [@Aprilhuu](https://github.com/Aprilhuu) [@lmxyy](https://github.com/lmxyy) just wondering if this is being actively worked on and/or will make it for the v0.3 release in late May? This issue now has the highest thumbs up count of all feature requests on the Nunchaku repo 🚀 If diffusers support is a prerequisite and is blocking this, then please let me know since I might be able to help speed that up, assuming you're waiting for an external contributor to solve that.
> 
> For those who are impatient and want to experiment, I've uploaded some unofficial, possibly low-quality/buggy weights here, based on [the script I mentioned earlier](https://gist.github.com/josephrocca/dd339fc20c18dd48524c57b6e4005486). Be sure to read the usage steps since the weights are Chroma "pretending to be" Schnell:
> 
> * https://huggingface.co/rocca/chroma-nunchaku-test

Thank you for your contribution. I don't know if it's because your model disguised it as FLUX, but it backfired. The chroma you quantified actually supports flux CN 2.0. I feel that this idea can be continued, especially when chroma originally cannot support it. Your model is obviously more worthy of use after it supports CN. In addition, the V36 detailed repair version has also been released. Compared with the previous versions, the quality of the latest version has been improved and the steps has also decreased. Although FB CACHE cannot be used at present, the serial connection of chroma 8s hyper lora can still increase the speed a lot.  Finally, the most exciting thing is that the VRAM usage is very low. I believe that the directions you are doing will greatly promote the ecology of the future community chroma (in addition, English is not my native language. I directly translated it with translation software and replied. Please forgive me if there are grammatical ambiguities)

**Comment by Ednaordinary** [comment]
Chroma PR is pretty close to done in https://github.com/huggingface/diffusers/pull/11698
I believe Nunchaku support should be easy from there? @Aprilhuu 

Take note of https://github.com/huggingface/diffusers/pull/11698#issuecomment-2969717180: distilled_guidance_layer may or may not play well with quantization. deepcompressor might be fine though

**Comment by ghost** [comment]
Thank you so much for working on this @Ednaordinary! 
I considered mentioning this yesterday in case you might have missed it but since you brought it up then I'll say it now that there is indeed some quality loss when some specific layers are quantized.
The guy maitining the FP8 Scaled uploads on HF recently started releasing a '_nodistill' version which skips the quantization of `distilled_guidance_layer, final_layer, img_in and txt_in` but apart from distilled_guidance_layer I'm not sure if its really necessary to skip the others. [Source](https://huggingface.co/Clybius/Chroma-fp8-scaled/discussions/8).
Saying this just in case that they should also be skipped.

**Comment by nitinmukesh** [comment]
Many thank to @Ednaordinary and diffusers team. The PR is now merged
https://github.com/huggingface/diffusers/pull/11698

@Aprilhuu @lmxyy 

Please if Chroma model integration in nunchaku can be considered.

**Comment by iamwavecut** [comment]
Just saying thanks to all you here pushing this to be implemented.

Chroma is a thing and if it would be supported in nunchaku, that may cause a lot of positive attention for the latter.

**Comment by AmericanPresidentJimmyCarter** [comment]
The current implementation seems broken: https://github.com/huggingface/diffusers/issues/11724

So you might want to wait before thinking about using nunchaku quant.

**Comment by Ednaordinary** [comment]
Quantization shouldn't rely on the attention mask being correct (unless training, which I don't believe this quantization method does)

**Comment by AmericanPresidentJimmyCarter** [comment]
iirc uses dataset based calibration with samples from the model so that you can tell if your configuration worked for the model or not: https://github.com/mit-han-lab/deepcompressor/tree/main/examples/diffusion

**Comment by Ednaordinary** [comment]
Ah good to know. I'll work on getting the attention mask working as soon as possible

**Comment by AmericanPresidentJimmyCarter** [comment]
Should be fixed now

**Comment by josephrocca** [comment]
CFG-baked checkpoint - i.e. use CFG=1 for 2x speedup: https://huggingface.co/lodestones/chroma-debug-development-only/resolve/main/staging_cfg_3/2025-06-28_06-54-17.pth This is based on v40, for reference.

I've tested it and it works great. It's baked at cfg=5.

@Aprilhuu IIRC, the deepcompressor config file already lets us set CFG value - if that's true, then I think we don't need any special deepcompressor config flag to support quantizing CFG-baked Chroma checkpoints?

**Comment by wbbzczc** [comment]
> CFG-baked checkpoint - i.e. use CFG=1 for 2x speedup: https://huggingface.co/lodestones/chroma-debug-development-only/resolve/main/staging_cfg_3/2025-06-28_06-54-17.pth This is based on v40, for reference.
> 
> I've tested it and it works great. It's baked at cfg=5.
> 
> [@Aprilhuu](https://github.com/Aprilhuu) IIRC, the deepcompressor config file already lets us set CFG value - if that's true, then I think we don't need any special deepcompressor config flag to support quantizing CFG-baked Chroma checkpoints?

Thanks for working out, so is this the chroma-unlocked-v40-detail-calibrated for nunchaku?

**Comment by DreamGenX** [comment]
@Aprilhuu Hi! Is this feature still something you plan to work on?

**Comment by Aprilhuu** [comment]
> [@Aprilhuu](https://github.com/Aprilhuu) Hi! Is this feature still something you plan to work on?

Hey DreamGenX, I am not actively working on this feature right now

**Comment by iamwavecut** [comment]
That's upsetting 😓

**Comment by ghost** [comment]
It seems the last version of Chroma (v50) has been released today.
We already have diffusers support so now its the best time to get nunchaku for it. 
I'd do it if I was smart enough. Can only wait for the AI overlords. 🚀

**Comment by herbtarlac** [comment]
With Nunchaku Qwen dropping I thought I'd check in on Chroma after v50.

**Comment by iamwavecut** [comment]
Any buzz?

**Comment by Stealt91** [comment]
Chroma is now out of development as Chroma1 release (as i understand it). I hope it can get support!

**Comment by Tablaski** [comment]
Desperate for a Chroma final version Nunchaku quant as well !

**Comment by iamwavecut** [comment]
🥲

**Comment by fappaz** [comment]
Just a reminder that soon™ we'll have an official release of Chroma Radiance [[1](https://huggingface.co/lodestones/chroma-debug-development-only/tree/main/radiance)][[2](https://www.reddit.com/r/StableDiffusion/comments/1mxwr4e/update_chroma_project_training_is_finished_the/)], a version that  operates directly in pixel space (dismissing VAE), improving detail fidelity while keeping same performance (according to Lodestone themself). This might become a big thing amongst image models, so it would be awesome if Nunchaku could support Chroma.

**Comment by Poltergei** [comment]
Don't hold your breathe, it's looking like never..
